src/Core/Managed/Shared/Extensibility/MetricManager.cs

```diff
@@ -0,0 +1,285 @@
+        /// with this instance of <see cref="MetricManager"/>.
+        /// </summary>
+        internal IList<IMetricProcessor> MetricProcessors
+        {
+            get
+            {
+                TelemetryConfiguration config = this.telemetryClient.TelemetryConfiguration;
+
+                return config.MetricProcessors;
+            }
+        }
+
+        /// <summary>
+        /// Creates metric.
+        /// </summary>
+        /// <param name="name">Name of the metric.</param>
+        /// <param name="dimensions">Optional dimensions.</param>
+        /// <returns>Metric instance.</returns>
+        public Metric CreateMetric(string name, IDictionary<string, string> dimensions = null)
+        {
+            if (this.disposed)
+            {
+                throw new ObjectDisposedException(nameof(MetricManager));
+            }
+
+            return new Metric(this, name, dimensions);
+        }
+
+        /// <summary>
+        /// Flushes the in-memory aggregation buffers.
+        /// </summary>
+        public void Flush()
+        {
+            if (this.disposed)
+            {
+                throw new ObjectDisposedException(nameof(MetricManager));
+            }
+
+            this.Flush(false);
+        }
+
+        /// <summary>
+        /// Disposes the object.
+        /// </summary>
+        public void Dispose()
+        {
+            this.disposed = true;
+
+            if (this.snapshotTimer != null)
+            {
+                this.snapshotTimer.Dispose();
+                this.snapshotTimer = null;
+            }
+
+            this.Flush(true);
+        }
+
+        internal SimpleMetricStatisticsAggregator GetStatisticsAggregator(Metric metric)
+        {
+            if (this.disposed)
+            {
+                throw new ObjectDisposedException(nameof(MetricManager));
+            }
+
+            return this.metricDictionary.GetOrAdd(metric, (m) => { return new SimpleMetricStatisticsAggregator(); });
+        }
+
+        /// <summary>
+        /// Calculates wait time until next snapshot of the aggregators.
+        /// </summary>
+        /// <returns>Wait time.</returns>
+        private static TimeSpan GetWaitTime()
+        {
+            DateTimeOffset currentTime = DateTimeOffset.UtcNow;
+
+            double minutesFromZero = currentTime.Subtract(DateTimeOffset.MinValue).TotalMinutes;
+
+            // we want to wake up exactly at 1 second past minute
+            // to make perceived system latency look smaller
+            var nextWakeTime = DateTimeOffset.MinValue
+                .AddMinutes((long)minutesFromZero)
+                .Add(snapshotFrequency)
+                .AddSeconds(1);
+
+            TimeSpan sleepTime = nextWakeTime - DateTimeOffset.UtcNow;
+
+            // adjust wait time to a bit longer than a minute if the wake up time is within few seconds from now
+            return sleepTime < TimeSpan.FromSeconds(3) ? sleepTime.Add(snapshotFrequency) : sleepTime;
```

Comment:
You are not adding a minute, but the period. NOt sure what's intended, but please make the comment and code match.